uitools/widgets/Esri/ArcGISRuntime/Toolkit/BookmarksView.cpp

```diff
@@ -0,0 +1,116 @@
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ ******************************************************************************/
+#include "BookmarksView.h"
+
+#include "ui_BookmarksView.h"
+
+// Toolkit headers
+#include "BookmarksViewController.h"
+#include "BookmarkListItem.h"
+
+// ArcGISRuntime headers
+#include <MapGraphicsView.h>
+#include <SceneGraphicsView.h>
+
+namespace Esri {
+namespace ArcGISRuntime {
+namespace Toolkit {
+
+  /*!
+  \class Esri::ArcGISRuntime::Toolkit::BookmarksView
+  \inmodule EsriArcGISRuntimeToolkit
+  \ingroup ArcGISQtToolkitUiCppWidgetsViews
+  \since Esri.ArcGISRuntime 100.15
+  \brief The user interface for the BookmarksView.
+  The BookmarksView displays a collection of bookmarks in the form of viewpoints from either ArcGIS Online, a user-defined portal,
+  or an array of Bookmarks. When the user selects a bookmark from the BookmarksView, the viewpoint rendered in the current
+  geoView is repalced in the given map/scene by the viewpoint selected from the bookmarks.
+
+  \note By default, the BookmarksView will attempt to fetch the set of developer bookmarks, which require an \l{https://developers.arcgis.com/qt/get-started/#3-access-services-and-content-with-an-api-key}{API key} to access.
+  */
+
+  /*!
+  \brief Default constructor.
+  \list
+    \li \a parent Parent widget.
+  \endlist
+
+  View mantains its associated controller, sets up the view itself and its model.
+  \note If this constructor is used, a \c GeoView must be set separately using \l setGeoModel.
+  */
+  BookmarksView::BookmarksView(QWidget* parent) :
+    QFrame(parent),
+    m_controller(new BookmarksViewController(this)),
+    m_ui(new Ui::BookmarksView)
+  {
+    m_ui->setupUi(this);
+    m_ui->listView->setModel(m_controller->bookmarks());
+
+    connect(m_ui->listView->selectionModel(), &QItemSelectionModel::currentChanged, this, &BookmarksView::onItemSelected);
+  }
+
+  /*!
+    \brief Destructor.
+   */
+  BookmarksView::~BookmarksView()
+  {
+    delete m_ui;
+  }
+
+  /*!
+    \brief Returns the controller.
+   */
+  BookmarksViewController* BookmarksView::controller() const
+  {
+    return m_controller;
+  }
+
+  /*!
+  \brief Set the \c GeoView.
+  \list
+    \li \a mapView Sets the \c GeoView to a \c MapView.
+  \endlist
+ */
+  void BookmarksView::setMapView(MapGraphicsView* mapView)
+  {
+    m_controller->setGeoView(mapView);
+  }
+
+  /*!
+  \brief Set the \c GeoView.
+  \list
+    \li \a sceneView Sets the \c GeoView to a \c SceneView.
+  \endlist
+ */
+  void BookmarksView::setSceneView(SceneGraphicsView* sceneView)
+  {
+    m_controller->setGeoView(sceneView);
+  }
+
+  /*!
+  \internal
+  \brief Slot that sets the current viewpoint with \a index.
+  Once linked to the clicked ListView event, this slot receives \c QModelIndex \a index and uses it
+  to set its viewpoint into the controller.
```

Comment:
🙃 When a click event is received, the bookmark at the given index is zoomed to via the controller.